backend/src/main/java/com/tourneynizer/tourneynizer/model/MatchGenerator.java

```diff
@@ -0,0 +1,90 @@
+package com.tourneynizer.tourneynizer.model;
+
+import com.tourneynizer.tourneynizer.dao.MatchDao;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+
+public class MatchGenerator {
+
+    private final MatchDao matchDao;
+
+    public MatchGenerator(MatchDao matchDao) {
+        this.matchDao = matchDao;
+    }
+
+    public List<Match> createTournamentMatches(List<Team> teams, User user, Tournament tournament) throws SQLException {
+        if (TournamentType.VOLLEYBALL_BRACKET.equals(tournament.getType())) {
+            return createBracketMatches(teams, tournament, user);
+        }
+        else {
+            throw new IllegalArgumentException(tournament.getType().name() + " is not currently supported");
+        }
+    }
+
+    private class MatchNode implements Comparable<MatchNode> {
+        private Match match;
+        private Team team;
+        public int orderInserted;
+
+        public MatchNode(Team team, int orderInserted, int height) {
+            this.setValue(team);
+            this.orderInserted = orderInserted;
+            this.height = height;
+        }
+
+        public MatchNode(Match match, int orderInserted, int height) {
+            this.setValue(match);
+            this.orderInserted = orderInserted;
+            this.height = height;
+        }
+
+        public void setValue(Match match) {
+            this.match = match;
+            this.team = null;
+        }
+
+        public void setValue(Team team) {
+            this.team = team;
+            this.match = null;
+        }
+
+        public long getValueId() {
+            return match != null ? match.getId() : team.getId();
+        }
+
+        public MatchNode child1, child2;
+        public int height;
+
+        @Override
+        public int compareTo(MatchNode o) {
+            int diff = this.height - o.height;
+            if (diff != 0) { return diff; }
+            return this.orderInserted - o.orderInserted;
+        }
+    }
+    private List<Match> createBracketMatches(List<Team> teams, Tournament tournament, User user) throws SQLException {
+        PriorityQueue<MatchNode> tree = new PriorityQueue<>();
+        int orderInserted = 0;
+        for (Team team : teams) {
+            tree.add(new MatchNode(team, orderInserted++, 0));
+        }
+
+        int order = 0;
+        List<Match> matches = new ArrayList<>();
+
+        while (tree.size() >= 2) {
+            MatchNode node1 = tree.poll();
+            MatchNode node2 = tree.poll();
+
+            Match parent = new Match(tournament.getId(), order++, node1.getValueId(), node2.getValueId(), null, ScoreType.ONE_SET);
+            matchDao.insert(parent, user);
+            tree.add(new MatchNode(parent, orderInserted++, Math.max(node1.height, node2.height) + 1));
+            matches.add(parent);
```

Comment:
Ok this looks pretty cool, but are you also prepared for a consolation bracket? Or is it just easier to do a one and done?

Reply:
This is easier; it's just a starting point.